### DISCUSSION TITLE
Blog on using Matter.js to control the brightness of a light (Part 2 of a previous blog)

### DISCUSSION THREAD

**Discussion by potto216** [matter-js/matter.js#1410] [Show and tell]
Hi All, I wrote a second blog on using Matter.js focusing on how to control the brightness of a light https://medium.com/@potto_94870/matter-js-adding-a-brightness-control-to-a-light-456513ed8ca0. The article should be accessible by anyone without a Medium account. I appreciate suggestions and corrections.

**Comment by Apollon77** [comment]
Cool,

some notes:

* When you "simplify" you use "npm init @matter" but without any example specified? especially because you want todo a light you could select the light example or such?
* Also: "npm init" automatically builds anything, so the build command is only needed if you change something. 
* In your code example when you convert to DimmableLightDevice ... some comments still say "OnOffLight" :-)
* "features like brightness can easily be added in nonstandard devices by including the Level Control cluster in an endpoint." is indeed difficult because it is formally undefined what happens if extra clusters exist on a device type. So likely most controllers and their UIs would simply ignore them and so the feature is "only half added" :-)
* when showing the code on root endpoint behaviors maybe add a sentence that most of the added Behaviors are "official clusters", but some (like Index, Parts" are matter.js-internal management behaviors that contain more basic logic for a "server node" and manage internal data structures

**Reply by potto216** [reply]
Great feedback!

- I used  `npm init @matter` to follow the bootstrapping a new project example in the readme. I see your point about specifying the project type especially if the initial project changes.
- I'll fix the init/build functionality issue
- I'll make changes to the code block with OnOffLightDevice in the comments.
- What is the best way for someone who is building a device that has multiple switches and multiple sensors to make a Matter interface? Would it be to look in the Matter Device Library and add each sensor as a different endpoint on a single node and possibly an Aggregator if they are related? Is there a generic multiple value device like the Generic Switch?
- thanks, will do.

**Reply by Apollon77** [reply]
For "Multiple" things but "fixed" on a device you normally build a so called "Composed Device".
That means you have one ServerNode and simply add different Endpoints to it (e.g. I have a Meross PlugIn Strip and this has simply 5x OnOffPlugInDevice there. but you could also build a multi sensor device with a TemperatureSensorDevice and a HumiditySensorDevice.
If you add multiple devices of the same type there (e.g. the 5x PlugIn example above) you are required to use "tagList" per endpoint to differentiate them. (see https://github.com/project-chip/matter.js/blob/main/chip-testing/src/AllClustersTestInstance.ts#L697-L704)

An Aggregator you usually use when you have "Non Matter devices" that you want to bring into the matter world where also the number of devices can differentiate and such ... thats an other use case

**Reply by Apollon77** [reply]
just tell when blog is final then I can plan posting on linked In again

**Reply by potto216** [reply]
Besides the npm init/build and code comment changes I made the following edits to https://medium.com/@potto_94870/matter-js-adding-a-brightness-control-to-a-light-456513ed8ca0

**Changed:**
"Figure 4 also illustrates the beauty of Matter's modular design - features like brightness can easily be added in nonstandard devices by including the Level Control cluster in an endpoint."
**to**
"Figure 4 also illustrates the beauty of Matter's modular design - features like brightness are added in devices defined in the Matter Device Library Specification by including the Level Control cluster in an endpoint."

**Changed:**
"These clusters define the administrative and diagnostic capabilities of the node, which are essential for device management such as commissioning. "
**to**
"These clusters define the administrative and diagnostic capabilities of the node, which are essential for device management such as commissioning. Most of the added Behaviors in the above code are "official clusters", but some like `Index` and `Parts` are matter.js-internal management behaviors that contain more basic logic for a "server node" and manage internal data structures."

**Changed:**
"And once your device works with Matter it works with any Matter compatible app such as the Google Home app, Amazon's Alexa app, or Apple's Home app, just to name a few."
**to**
"And once your device works with Matter it works with any Matter compatible app such as the Google Home app, Amazon's Alexa app, or Apple's Home app, just to name a few. But this compatibility requires the device to be defined in the Matter Device Library Specification. What if it isn't? What if the device has extra sensors or switches? One answer to this is the use of Composed Devices which will be the subject of another article."

Let me know of other change suggestions.

**Reply by Apollon77** [reply]
Awesome!

**Reply by Apollon77** [reply]
https://www.linkedin.com/posts/ifischer_matterjs-adding-a-brightness-control-to-activity-7266721859436527616-HXml?utm_source=share&utm_medium=member_desktop
